core/src/main/scala/org/apache/spark/Accumulators.scala

```diff
@@ -307,11 +310,22 @@ private[spark] object Accumulators {
     }
   }
 
+  def remove(accId : Long) {
+    synchronized {
+      originals.remove(accId)  
+    }
+  }
+  
   // Get the values of the local accumulators for the current thread (by ID)
   def values: Map[Long, Any] = synchronized {
     val ret = Map[Long, Any]()
     for ((id, accum) <- localAccums.get) {
-      ret(id) = accum.localValue
+      // Since we are now storing weak references, we must check whether the underlying data
+      // is valid. 
+      ret(id) = accum.get match {
+        case Some(values) => values.localValue
+        case None => None
```

Comment:
This was dumb for me to overlook, too: if the data structure is invalid, then this would just silently ignore it; even if this was a rare error-condition, there should have been a warning here.